### PR TITLE
[WIP] Refactor SBOM scanner

### PR DESCRIPTION
**Pull request by AliDatadog**
### Reviewer's Checklist

- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [ ] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [ ] Adequate QA/testing plan information is provided. Except if the `qa/skip-qa` label, with required either `qa/done` or `qa/no-code-change` labels, are applied.
- [ ] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the `k8s/<min-version>` label, indicating the lowest Kubernetes version compatible with this feature.
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

### CODE REVIEW

File: comp/core/workloadmeta/collectors/internal/containerd/image_sbom_trivy.go [hunk 1]
```diff
@@ -45,7 +45,14 @@ func (c *collector) startSBOMCollection(ctx context.Context) error {
 		workloadmeta.NormalPriority,
 		workloadmeta.NewFilter(&filterParams),
 	)
-	resultChan := make(chan sbom.ScanResult, 2000)
+	scanner := collectors.GetContainerdScanner()
+	if scanner == nil {
+		return fmt.Errorf("error retrieving global docker scanner")
```
Comment (clamoriniere) [inline comp/core/workloadmeta/collectors/internal/containerd/image_sbom_trivy.go:50]: you are mentioning docker but it is a containerd scanner

```suggestion
		return fmt.Errorf("error retrieving global containerd scanner")
```

Comment (clamoriniere) [reply comp/core/workloadmeta/collectors/internal/containerd/image_sbom_trivy.go:50]: is it possible that `scanner` is `nil`? 

can we split this function in two:
* `initSBOMCollection()` responsible of setting the instances of `c.sbomScanner` and `scanner`, setting the result channel  and call this function in `NewCollector()`
* `startSBOMCollection()` responsible to start the collection (register to workload) start the go routine.

File: comp/core/workloadmeta/collectors/internal/containerd/image_sbom_trivy.go [hunk 2]
```diff
@@ -45,7 +45,14 @@ func (c *collector) startSBOMCollection(ctx context.Context) error {
...
+	}
+	resultChan := scanner.Channel()
```
Comment (clamoriniere) [inline comp/core/workloadmeta/collectors/internal/containerd/image_sbom_trivy.go:52]: `resultChan` is not use in this function except to check it is `nil`

File: comp/core/workloadmeta/collectors/internal/docker/image_sbom_trivy.go [hunk 3]
```diff
@@ -53,7 +52,14 @@ func (c *collector) startSBOMCollection(ctx context.Context) error {
...
 
-	resultChan := make(chan sbom.ScanResult, 2000)
+	scanner := collectors.GetDockerScanner()
+	if scanner == nil {
+		return fmt.Errorf("error retrieving global docker scanner")
+	}
+	resultChan := scanner.Channel()
+	if resultChan == nil {
+		return fmt.Errorf("error retrieving global docker scanner channel")
+	}
```
Comment (clamoriniere) [inline comp/core/workloadmeta/collectors/internal/docker/image_sbom_trivy.go:62]: not use in this function

File: comp/core/workloadmeta/collectors/internal/containerd/image_sbom_trivy.go [hunk 4]
```diff
@@ -45,7 +45,14 @@ func (c *collector) startSBOMCollection(ctx context.Context) error {
...
+	if resultChan == nil {
+		return fmt.Errorf("error retrieving global docker scanner channel")
```
Comment (clamoriniere) [inline comp/core/workloadmeta/collectors/internal/containerd/image_sbom_trivy.go:54]:
```suggestion
		return fmt.Errorf("error retrieving global containerd scanner channel")
```

File: comp/core/workloadmeta/collectors/internal/docker/image_sbom_trivy.go [hunk 5]
```diff
@@ -86,61 +92,65 @@ func (c *collector) startSBOMCollection(ctx context.Context) error {
...
 					status = workloadmeta.Failed
 					reportedError = result.Error.Error()
+				} else {
+					bom, err := result.Report.ToCycloneDX()
+					if err != nil {
+						log.Errorf("Failed to extract SBOM from report")
+						status = workloadmeta.Failed
+						reportedError = result.Error.Error()
+					}
+					report = bom
```
Comment (clamoriniere) [inline comp/core/workloadmeta/collectors/internal/docker/image_sbom_trivy.go:120]: for clarity this part can be extracted in a function.

Comment (clamoriniere) [reply comp/core/workloadmeta/collectors/internal/docker/image_sbom_trivy.go:120]: same for the previous `for loop` line 64 https://github.com/DataDog/datadog-agent/pull/22639/files#diff-84880b15ffd9ed1c97f5a97eed3b3bc24504f53ea426125b6925e89ecc0c6f8eR64

**Review by clamoriniere — COMMENTED** [review]
started the review

File: comp/core/workloadmeta/collectors/internal/containerd/image_sbom_trivy.go (hunk 2 above)
Comment (AliDatadog) [reply comp/core/workloadmeta/collectors/internal/containerd/image_sbom_trivy.go:52]: I pass it down here
https://github.com/DataDog/datadog-agent/blob/098c1857e3067db961d613d79a4f9e4d75dbb53f/comp/core/workloadmeta/collectors/internal/containerd/image_sbom_trivy.go#L72

We should be able to call `scanner.Channel()` directly in the function `startScanResultHandler`. However I think that if we want to add some unit tests in the future, we'll need mocks and passing using `global` state should be avoided where we can.
WDYT ?

File: comp/core/workloadmeta/collectors/internal/docker/image_sbom_trivy.go (hunk 3 above)
Comment (AliDatadog) [reply comp/core/workloadmeta/collectors/internal/docker/image_sbom_trivy.go:62]: Same as above, we use it here:
https://github.com/DataDog/datadog-agent/blob/098c1857e3067db961d613d79a4f9e4d75dbb53f/comp/core/workloadmeta/collectors/internal/docker/image_sbom_trivy.go#L97

Maybe we should have some more generic code between docker and containerd but I leave it for another PR.

**Pull request closed without merging**